File: clients/python/games/mastermind.py
from collections import Counter
from typing import List, Optional
from games.base import BaseGame
from framework.i18n import t
import json


class Mastermind(BaseGame):
    MAX_GUESSES = 10

    def __init__(self):
        self.players: List[str] = []
        self._code: Optional[List[int]] = None
        self._guesses: List = []
        self._turn_idx = 0
        self._over = False
        self._winner = None

    def start(self, players: List[str]) -> None:
        self.players = players
        self._code = None
        self._guesses = []
        self._turn_idx = 0
        self._over = False
        self._winner = None

    def current_turn(self) -> Optional[str]:
        return self.players[self._turn_idx] if self.players else None

    def validate_move(self, player_id: str, move_data: dict) -> bool:
        if self._over or player_id != self.current_turn():
            return False
        if self._code is None:
            code = move_data.get('code', [])
            return len(code) == 4 and all(isinstance(d, int) and 1 <= d <= 6 for d in code)
        guess = move_data.get('guess', [])
        return len(guess) == 4 and all(isinstance(d, int) and 1 <= d <= 6 for d in guess)

    def apply_move(self, player_id: str, move_data: dict) -> None:
        if self._code is None:
            self._code = move_data['code']
            self._turn_idx = 1
        else:
            guess = move_data['guess']
            exact, mis = self._score(guess)
            self._guesses.append((guess, exact, mis))
            if exact == 4:
                self._over = True
                self._winner = self.players[1]
            elif len(self._guesses) >= self.MAX_GUESSES:
                self._over = True
                self._winner = self.players[0]

    def _score(self, guess):
        exact = sum(g == c for g, c in zip(guess, self._code))
        mis = sum((Counter(self._code) & Counter(guess)).values()) - exact
        return exact, mis

    def is_over(self):
        return self._over, self._winner

    def render(self, perspective: Optional[str] = None) -> str:
        lines = [t('mastermind.title')]
        for i, (g, e, m) in enumerate(self._guesses):
            lines.append(t('mastermind.guess', n=i + 1, guess=g, exact=e, mis=m))
        if not self._over:
            lines.append(t('mastermind.remaining', n=self.MAX_GUESSES - len(self._guesses)))
        return '\n'.join(lines)

    def get_state(self, perspective: Optional[str] = None) -> dict:
        state = {'guesses': self._guesses, 'turn': self.current_turn(),
                 'players': self.players, 'over': self._over}
        if perspective == self.players[0] or self._over:
            state['code'] = self._code
        return state

    def load_state(self, data: dict, perspective: Optional[str] = None) -> None:
        if not data:
            return
        if 'players' in data:
            self.players = list(data['players'])
        if 'guesses' in data:
            self._guesses = list(data['guesses'])
        if 'code' in data:
            self._code = data['code']
        if 'over' in data:
            self._over = bool(data['over'])
        if 'turn' in data and data['turn'] in self.players:
            self._turn_idx = self.players.index(data['turn'])

    def parse_input(self, raw: str) -> Optional[dict]:
        raw = raw.strip()
        if raw.startswith(('{', '[')):
            try:
                obj = json.loads(raw)
                if isinstance(obj, dict):
                    return obj
            except ValueError:
                pass
        parts = raw.split()
        digits = None
        if len(parts) == 4:
            try:
                digits = [int(p) for p in parts]
            except ValueError:
                return None
        elif len(parts) == 1 and len(parts[0]) == 4 and parts[0].isdigit():
            digits = [int(c) for c in parts[0]]
        if digits is not None and len(digits) == 4:
            return {'code': digits} if self._code is None else {'guess': digits}
        return None

    def get_help(self) -> List[str]:
        return [
            'Guess the secret 4-digit code (digits 1-6).',
            'B = right digit + right position.  W = right digit, wrong position.',
            'Move: <d1> <d2> <d3> <d4>   e.g. "1 2 3 4"  or  "1234"',
        ]

File: clients/node/src/games/mastermind.js
'use strict';
const { BaseGame } = require('./base');
const { t } = require('../framework/i18n');

const MAX_GUESSES = 10;

class Mastermind extends BaseGame {
  constructor() {
    super();
    this.players = [];
    this._code = null;
    this._guesses = [];
    this._turnIdx = 0;
    this._over = false;
    this._winner = null;
  }

  start(players) {
    this.players = players;
    this._turnIdx = 0;
  }

  currentTurn() {
    return this.players[this._turnIdx] || null;
  }

  validateMove(playerId, moveData) {
    if (this._over || playerId !== this.currentTurn()) return false;
    if (this._code === null) {
      const code = moveData.code || [];
      return code.length === 4 && code.every(d => Number.isInteger(d) && d >= 1 && d <= 6);
    } else {
      const guess = moveData.guess || [];
      return guess.length === 4 && guess.every(d => Number.isInteger(d) && d >= 1 && d <= 6);
    }
  }

  applyMove(playerId, moveData) {
    if (this._code === null) {
      this._code = moveData.code;
      this._turnIdx = 1;
    } else {
      const guess = moveData.guess;
      const [exact, mis] = this._score(guess);
      this._guesses.push([guess, exact, mis]);
      if (exact === 4) {
        this._over = true;
        this._winner = this.players[1];
      } else if (this._guesses.length >= MAX_GUESSES) {
        this._over = true;
        this._winner = this.players[0];
      }
    }
  }

  _score(guess) {
    let exact = 0;
    for (let i = 0; i < 4; i++) {
      if (guess[i] === this._code[i]) exact++;
    }
    const codeCount = {};
    const guessCount = {};
    for (let i = 0; i < 4; i++) {
      codeCount[this._code[i]] = (codeCount[this._code[i]] || 0) + 1;
      guessCount[guess[i]] = (guessCount[guess[i]] || 0) + 1;
    }
    let total = 0;
    for (const k of Object.keys(codeCount)) {
      total += Math.min(codeCount[k], guessCount[k] || 0);
    }
    return [exact, total - exact];
  }

  isOver() {
    return [this._over, this._winner];
  }

  render(perspective) {
    const lines = [t('mastermind.title')];
    for (let i = 0; i < this._guesses.length; i++) {
      const [g, e, m] = this._guesses[i];
      lines.push(t('mastermind.guess', { n: i + 1, guess: g, exact: e, mis: m }));
    }
    if (!this._over) {
      lines.push(t('mastermind.remaining', { n: MAX_GUESSES - this._guesses.length }));
    }
    return lines.join('\n');
  }

  getState(perspective) {
    const state = {
      guesses: this._guesses,
      turn: this.currentTurn(),
      players: this.players,
      over: this._over,
    };
    if (perspective === this.players[0] || this._over) {
      state.code = this._code;
    }
    return state;
  }

  loadState(data) {
    if (!data) return;
    if (data.players) this.players = data.players;
    if (data.guesses) this._guesses = data.guesses.map(g => [...g]);
    if (data.over != null) this._over = data.over;
    if (data.code) this._code = [...data.code];
    if (data.turn != null) {
      const idx = this.players.indexOf(data.turn);
      this._turnIdx = idx >= 0 ? idx : 0;
    }
  }
}

module.exports = { Mastermind };
